core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -1017,24 +1041,29 @@ public void onFailure(Exception shardFailedError) {
     }
 
     /** a wrapper class to encapsulate a request when being sent to a specific allocation id **/
-    public static final class ConcreteShardRequest<R extends TransportRequest> extends TransportRequest {
+    public static final class ConcreteShardRequest<R extends ReplicationRequest<?>> extends TransportRequest {
 
         /** {@link AllocationId#getId()} of the shard this request is sent to **/
         private String targetAllocationID;
 
+        private long primaryTerm;
+
         private R request;
 
         ConcreteShardRequest(Supplier<R> requestSupplier) {
             request = requestSupplier.get();
             // null now, but will be populated by reading from the streams
             targetAllocationID = null;
+            primaryTerm = 0L;
         }
 
-        ConcreteShardRequest(R request, String targetAllocationID) {
+        ConcreteShardRequest(R request, String targetAllocationID, long primaryTerm) {
             Objects.requireNonNull(request);
             Objects.requireNonNull(targetAllocationID);
             this.request = request;
             this.targetAllocationID = targetAllocationID;
+            this.primaryTerm = primaryTerm;
+            this.request.primaryTerm(primaryTerm); // for bwc with ES < v5.6, that has the primary term on the inner ReplicationRequest
```

Comment:
can we add an assertion to replication request that allows to set it once to a non zero value and makes sure that future callers keep it the same?